scrapy/utils/misc.py

```diff
@@ -137,17 +137,23 @@ def create_instance(objcls, settings, crawler, *args, **kwargs):
     ``*args`` and ``**kwargs`` are forwarded to the constructors.
 
     Raises ``ValueError`` if both ``settings`` and ``crawler`` are ``None``.
+
+    Raises ``TypeError`` if the resulting instance is ``None`` (e.g. if an
+    extension has not been implemented correctly).
     """
     if settings is None:
         if crawler is None:
             raise ValueError("Specify at least one of settings and crawler.")
         settings = crawler.settings
     if crawler and hasattr(objcls, 'from_crawler'):
-        return objcls.from_crawler(crawler, *args, **kwargs)
+        instance = objcls.from_crawler(crawler, *args, **kwargs)
     elif hasattr(objcls, 'from_settings'):
-        return objcls.from_settings(settings, *args, **kwargs)
+        instance = objcls.from_settings(settings, *args, **kwargs)
     else:
-        return objcls(*args, **kwargs)
+        instance = objcls(*args, **kwargs)
+    if instance is None:
+        raise TypeError("Instance is None for %s" % objcls)
```

Comment:
I think it may be useful to indicate which class method (`from_crawler` or `from_settings`) caused the issue. You could do something like:

```py
raise TypeError("%s.%s returned None" % (objcls.__qualname__, method_name))
```

Reply:
Done